### DISCUSSION TITLE
Remove --tw- variables from universal selector

### DISCUSSION THREAD

**Discussion by vascoeduardo** [tailwindlabs/tailwindcss#7317] [Help] [answered]
Hey everybody,

at the moment I'm facing a problem with tailwind variables that are declared in the universal selector *, :after, :before { }. The Problem is, that css variables in the universal selector seem to cause a super high CPU load on older iOS devices on every interaction(click, touch etc.). If I remove them from the universal selector, everything seems to be fine. My generated tailwind.css file has this universal selector:
```
*, :after, :before {
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x: ;
    --tw-pan-y: ;
    --tw-pinch-zoom: ;
    --tw-scroll-snap-strictness: proximity;
    --tw-ordinal: ;
    --tw-slashed-zero: ;
    --tw-numeric-figure: ;
    --tw-numeric-spacing: ;
    --tw-numeric-fraction: ;
    --tw-ring-inset: ;
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgb(59 130 246/0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur: ;
    --tw-brightness: ;
    --tw-contrast: ;
    --tw-grayscale: ;
    --tw-hue-rotate: ;
    --tw-invert: ;
    --tw-saturate: ;
    --tw-sepia: ;
    --tw-drop-shadow: ;
    --tw-backdrop-blur: ;
    --tw-backdrop-brightness: ;
    --tw-backdrop-contrast: ;
    --tw-backdrop-grayscale: ;
    --tw-backdrop-hue-rotate: ;
    --tw-backdrop-invert: ;
    --tw-backdrop-opacity: ;
    --tw-backdrop-saturate: ;
    --tw-backdrop-sepia: ;
}
```
Is it somehow possible to remove the variables declaration from the universal selector and move them to the specific class in which they are needed. I took a look at other tailwind.css files and it seems that it is somehow possible.

For example on tailwindcss.com only the  --tw-content variable is declared in the universal selector:
```
:after,:before {
    --tw-content: ""
}
```
To demonstrate what the problem on iOS devices , I screen captured the CPU usage of a iPad mini with tailwind variables declared in the universal selector. What you see is what happens if a empty div is touched:

https://user-images.githubusercontent.com/21193139/152379876-62705f1c-05cd-46c8-8907-ed360eec7ba2.mov

and here without the variables:

https://user-images.githubusercontent.com/21193139/152379887-46d90505-c4c0-4662-9486-c691e53fa4fe.mov

there is a huge difference in CPU usage so the website is barely usable because the interface is blocked for nearly 5 seconds

I hope somebody has an idea, what I am trying to achieve here.

**Comment by adamwathan** [comment, marked as answer]
Hey! Try this:

```js
// tailwind.config.js
module.exports = {
  experimental: {
    optimizeUniversalDefaults: true
  }
  // ...
}
```

We tried to make this strategy the default for v3 but it's not compatible with CSS modules or component-specific styles in Svelte/Vue components so we had to revert to using the universal stuff by default.

We use that option on the tailwindcss.com site though, and as long as you are using Tailwind the recommended way (as just a simple boring global stylesheet without CSS modules or any other scoped styles that cause PostCSS to run multiple times in isolation) you should be good.

**Reply by vascoeduardo** [reply]
Hey Adam,

thank you very much for your answer. 
I tried using your suggested solution and it seems to work properly. After thinking about it one more time I was thinking if it is somehow possible to declared the tailwind variables in the :root selector instead of declaring it in the  universal selector *. As we are using postcss-preset-env and with it postcss-custom-properties the variables would be copied to the desired position + we would have a fallback for browser wich don't support CSS Variables. What postcss-custom-properties does ist this:

```css
:root {
  --color: red;
}

h1 {
  color: var(--color);
}

/* becomes */

:root {
  --color: red;
}

h1 {
  color: red;
  color: var(--color);
}
```
Once again: thank you for your help!
Vasco

**Reply by adamwathan** [reply]
Hey! We can’t declare on root because then they would be inherited, and we need the values to start with their default for each element.

For example, if we set the defaults on root and one element had `ring-blue-500`, the ring variable wouldn’t be reset on any child elements so any child element setting any type of shadow would inherit the ring shadow which isn’t what you want.

**Reply by AndreasHeintze** [reply]
@adamwathan I had the same problem, so I did as you suggested and the variables where gone. I use Svelte, will there be a problem? Seems to work so far...

**Reply by mpalpha** [reply]
Perhaps making ```--tw-*``` variable defaults compatible with ```@layer``` or the ability to specify a custom selector could be a workable solution that would allow more flexibility for custom web components.

**Reply by wongjn** [reply]
The option does not remove all `--tw` but optimizes them down to used ones, see this [Tailwind Play example (open the *Generated CSS* panel)](https://play.tailwindcss.com/1MT3KaD5TN). Toggle the option to `false` in the *Config* tab to see the difference.

**Reply by djmtype** [reply]
@wongjn I had deleted my comment because I was wrong, and noticed that after the fact.

**Comment by bakura10** [comment]
Hi @adamwathan ,

I have tried this option, and I end up having some strange CSS, that I am not sure is expected. For instance, using the universal selector I end up with one selector like that:

<img width="613" alt="image" src="https://user-images.githubusercontent.com/1198915/164604270-fb143df0-30ef-428c-88c9-ca0cf1dc84b5.png">

And later another rule that simply seems to override it:

<img width="908" alt="image" src="https://user-images.githubusercontent.com/1198915/164604291-88c3733b-6d5f-4b3d-a337-c65b9f58953c.png">

I am also using some components beyond utility, and I end up having those kind of strange selectors:

<img width="737" alt="image" src="https://user-images.githubusercontent.com/1198915/164604350-8f500e08-5047-4a00-aa5a-6ce6934debd6.png">

Is this the expected output of this option?

I feel a way better approach here would be to try to implement that: https://github.com/tailwindlabs/tailwindcss/issues/8135 to strip down unused variables from both the generated selectors _and_ the universal selectors. I think that most people do not even use half of those variables.

A very naive approach that would simply state for instance "if no sepia filter class is used in the whole codebase, then sepia variable is not outputted" would work I think.

**Comment by danielimmke** [comment]
As a tag up to this, I recently experienced an issue where the `--tw-sepia` variable was causing a rendering issue in Webkit, specifically. I don't know if the previous comment mentioning the sepia filter specifically is related, but I am posting it on this issue for posterity.

To fix it, I had to manually add `--tw-sepia: unset` to the specific container div. I've attached a video of the issue I was experiencing.

https://github.com/user-attachments/assets/86ede315-ac9b-434a-a92d-8f06a7bf6430

**Comment by Inwerpsel** [comment]
Big universal rules are also pretty detrimental to the UX of dev tools, at least in chromium browsers. It shows up on every parent element of the selected element unnecessarily, completely striked through because that rule matches all lower elements. It makes the inspector view several times longer with pure garbage.

https://github.com/user-attachments/assets/0e3456db-29a9-4708-a7de-c72b534d2ad9

In a deep DOM this gets really bad and also grinds the performance of dev tools to a halt as it renders thousands of pointless elements.

**Reply by Inwerpsel** [reply]
Is there a way to avoid this behavior and still get a Styles panel without repetitions like this? It kind of baffles me that not more people are bothered by this, so I assume I'm missing something.

**Reply by djmtype** [reply]
Looking at your video, it appears like that's a production build rather than development, correct? If you're using a Vite project, in your config, try setting `css: { devSourcemap: true }`. Does that cut down any of the noise?

**Reply by Inwerpsel** [reply]
Yes, it's a production build, but not my own site, just the latest Tailwind site I came across (https://www.meetup.com/devs-gent/).

It's just an example though, for me all Tailwind sites with this rule look like this in the inspector.

I did try copy pasting the same CSS rule into my own project which does have source maps for CSS to test if it has any impact. It doesn't, the rule is still displayed in the same way, with the only difference being the file name linking to the actual source.

![twuniversalchaos](https://github.com/user-attachments/assets/f2ba55e3-c9bd-47b2-a3c4-6797ec195bf8)

@djmtype Do you have an example where a source map improves this behavior?

**Reply by adamwathan** [reply]
Nothing we can do about this on our end unfortunately, would have to be a setting in Chrome itself to control whether they show inherited styles in dev tools or something :/

Good news is that in v4 we use `@property` instead of the universal selector since we are only targeting Safari >= 16.4, so this stuff all goes away.

Check the Headless UI website as an example:

https://headlessui.com/

**Reply by adamwathan** [reply]
Seems like this causing performance issues might be a new thing because of an optimization they tried to make in Chrome:

https://github.com/tailwindlabs/tailwindcss/discussions/14713

Going to try to escalate this today with my contacts there.

**Reply by Inwerpsel** [reply]
In my earlier message I was mostly referring to slow performance just due to the sheer amount of elements, but it's indeed likely that there are performance issues when using universal selector and custom properties.

In the [Chromium commit](https://chromium-review.googlesource.com/c/chromium/src/+/5470982) linked in https://github.com/tailwindlabs/tailwindcss/discussions/14713, the following part of the description looks like it could be related.

> To correctly have a unique StyleVariables data for each style
introduces has_own_*_variables_ to keep track of the state.
(This is somewhat a bit manual which is unfortunate). This replaces
the HasOneRef call.

It seems very plausible that universal selectors don't play well with this change. If I'm reading it correctly, this should be true for every element once you start using a universal selector with custom properties. It's hard to tell from the Chromium code what would be the impact of this being true 100% of the time, but it feels like it probably breaks some optimization(s).

**Comment by djmtype** [comment]
@Inwerpsel PostCSS JIT Props might be worth trying. 

By example: https://stackblitz.com/edit/github-7wftaw

Honestly, I think Tailwind v4 would be your best bet even if it's still in alpha. It really depends on your project's browser support requirements. 

There's probably good reason why they were placed on the `*` selector originally due to the cascade, etc.

By adding JIT Props, you can place them within the `:root`, or back onto the `*, ::before, ::after`. The main difference is they would be opt in, so it should reduce the bloat. Just know, targeting `*, ::before, ::after`, could lead you back to square one. 

Install [PostCSS JIT Props](https://www.npmjs.com/package/postcss-jit-props).

Remove the `--tw-*` props entirely.
```js
// tailwind.config.cjs
experimental: {
    optimizeUniversalDefaults: true
}
```
Now add the `--tw-*` props back in, on demand.

Create a `postcss.config.cjs` and dump the `--tw-*` props in there. 

```js
module.exports = {
  plugins: [
    require('postcss-import'),
    require('tailwindcss'),
    require('postcss-jit-props')({
    '--tw-border-spacing-x': '0',
    '--tw-border-spacing-y': '0',
    '--tw-translate-x': '0',
    '--tw-translate-y': '0',
    '--tw-rotate': '0',
    '--tw-skew-x': '0',
    '--tw-skew-y': '0',
    '--tw-scale-x': '1',
    '--tw-scale-y': '1',
    '--tw-pan-x': ' ',
    '--tw-pan-y': ' ',
    '--tw-pinch-zoom': ' ',
    '--tw-scroll-snap-strictness': 'proximity',
    '--tw-gradient-from-position': ' ',
    '--tw-gradient-via-position': ' ',
    '--tw-gradient-to-position': ' ',
    '--tw-ordinal': ' ',
    '--tw-slashed-zero': ' ',
    '--tw-numeric-figure': ' ',
    '--tw-numeric-spacing': ' ',
    '--tw-numeric-fraction': ' ',
    '--tw-ring-inset': ' ',
    '--tw-ring-offset-width': '0px',
    '--tw-ring-offset-color': '#fff',
    '--tw-ring-color': 'rgb(59 130 246 / 0.5)',
    '--tw-ring-offset-shadow': '0 0 #0000',
    '--tw-ring-shadow': '0 0 #0000',
    '--tw-shadow': '0 0 #0000',
    '--tw-shadow-colored': '0 0 #0000',
    '--tw-blur': ' ',
    '--tw-brightness': ' ',
    '--tw-contrast': ' ',
    '--tw-grayscale': ' ',
    '--tw-hue-rotate': ' ',
    '--tw-invert': ' ',
    '--tw-saturate': ' ',
    '--tw-sepia': ' ',
    '--tw-drop-shadow': ' ',
    '--tw-backdrop-blur': ' ',
    '--tw-backdrop-brightness': ' ',
    '--tw-backdrop-contrast': ' ',
    '--tw-backdrop-grayscale': ' ',
    '--tw-backdrop-hue-rotate': ' ',
    '--tw-backdrop-invert': ' ',
    '--tw-backdrop-opacity': ' ',
    '--tw-backdrop-saturate': ' ',
    '--tw-backdrop-sepia': ' ',
    '--tw-contain-size': ' ',
    '--tw-contain-layout': ' ',
    '--tw-contain-paint': ' ',
    '--tw-contain-style': ' '
    }),
   require('autoprefixer')
  ]
}
```
